test/qa_scheduler.cpp

```diff
@@ -0,0 +1,222 @@
+template<typename T, std::int64_t N>
+class random_source2 : public fg::node<random_source2<T, N>, fg::OUT<T, 0, std::numeric_limits<std::size_t>::max(), "out">> {
+    trace_vector &tracer;
+    std::int64_t count = 0;
+public:
+    random_source2(trace_vector &trace, std::string_view name) : tracer{trace} {this->_name = name;}
+
+    constexpr std::int64_t
+    available_samples(const random_source2 &d) noexcept {
+        const auto ret = std::min(static_cast<std::int64_t>(N - count), std::numeric_limits<std::int64_t>::max());
+        return ret > 0 ? ret : -1; // '-1' -> DONE, produced enough samples
+    }
+
+    constexpr T
+    process_one() {
+        trace(tracer, this->name());
+        return static_cast<int>(1000 + count++);
+    }
+};
+
+template<typename T>
+class cout_sink : public fg::node<cout_sink<T>, fg::IN<T, 0, std::numeric_limits<std::size_t>::max(), "in">> {
+    trace_vector &tracer;
+    std::int64_t count = 0;
+public:
+    cout_sink(trace_vector &trace, std::string_view name) : tracer{trace} {this->_name = name;}
+    [[nodiscard]] fg::work_return_t
+    process_bulk(std::span<const T> input) noexcept {
+        trace(tracer, this->name(), static_cast<size_t>(input.size()));
+        fmt::print("data[{}]: data[0] = {}\n", input.size(), input[0]);
+        for (auto data: input) {
+            boost::ut::expect(boost::ut::that % data == 2 * count + 1000 + count);
+            count++;
+        }
+        return fg::work_return_t::OK;
+    }
+    constexpr void
+    process_one(T /*a*/) const noexcept {
+        trace(tracer, this->name());
+    }
+};
+
+template<typename T, T Scale, typename R = decltype(std::declval<T>() * std::declval<T>())>
+class scale : public fg::node<scale<T, Scale, R>, fg::IN<T, 0, std::numeric_limits<std::size_t>::max(), "original">, fg::OUT<R, 0, std::numeric_limits<std::size_t>::max(), "scaled">> {
+    trace_vector &tracer;
+public:
+    scale(trace_vector &trace, std::string_view name) : tracer{trace} {this->_name = name;}
+    template<fair::meta::t_or_simd<T> V>
+    [[nodiscard]] constexpr auto
+    process_one(V a) const noexcept {
+        trace(tracer, this->name(), vir::stdx::is_simd_v<V> ? V::size() : 1);
+        return a * Scale;
+    }
+};
+
+template<typename T, typename R = decltype(std::declval<T>() + std::declval<T>())>
+class adder : public fg::node<adder<T>, fg::IN<T, 0, std::numeric_limits<std::size_t>::max(), "addend0">, fg::IN<T, 0, std::numeric_limits<std::size_t>::max(), "addend1">, fg::OUT<R, 0, std::numeric_limits<std::size_t>::max(), "sum">> {
+    trace_vector &tracer;
+public:
+    adder(trace_vector &trace, std::string_view name) : tracer{trace} {this->_name = name;}
+    template<fair::meta::t_or_simd<T> V>
+    [[nodiscard]] constexpr auto
+    process_one(V a, V b) const noexcept {
+        trace(tracer, this->name(), vir::stdx::is_simd_v<V> ? V::size() : 1);
+        return a + b;
+    }
+};
+
+/**
```

Comment:
very nice test!

please add two very simple tests verifying the execution order for the following graph topologies for both the simple and BFS based scheduler:

 a) source -> node 1 -> node 2 -> sink result: both scheduler should produce the order src->1->2->sink
 b) split paths:
     src -> node 1 -> node 2 -> sink 1
     src -> node 3 -> node 4 -> sink 2
     
      expected results:
      simple: src->1->2->sink1->3->4->sink2
      BFS: src->1->3->2->4->sink1->sink2
      
That should conclude the testing for these basic types of scheduler ... more advanced scheduling should be done in further PRs.